### PR TITLE
Syscheck improvements (distinct) and date_field fixes

### PR DESCRIPTION
**Pull request by davidjiglesias**
|Related issue|
|---|
|#4466|

## Description

This PR adds changes to WazuhDbQuery to account for distinct parameter and fixes date_fields errors.

## Tests

All unit tests updated and passed:
`========================== 638 passed in 2.53 seconds ==========================`

### CODE REVIEW

**Comment by wazuhci** [comment]
The stress test has been launched. The charts will be posted when they are ready.
I will use the following configuration:
### General
- Commit: 1a68474
- Packages: 
  - [manager 3.12.0 0.branchdev4466syscheckendpointimprovements.commit1a68474 rpm](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/rpm/var/wazuh-manager-3.12.0-0.branchdev4466syscheckendpointimprovements.commit1a68474.x86_64.rpm)
  - [agent 3.12.0 0.branchdev4466syscheckendpointimprovements.commit1a68474 rpm](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/rpm/var/wazuh-agent-3.12.0-0.branchdev4466syscheckendpointimprovements.commit1a68474.x86_64.rpm)
  - [manager 3.12.0 0.branchdev4466syscheckendpointimprovements.commit1a68474 deb](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/deb/var/wazuh-manager_3.12.0-0.branchdev4466syscheckendpointimprovements.commit1a68474_amd64.deb)
  - [agent 3.12.0 0.branchdev4466syscheckendpointimprovements.commit1a68474 deb](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/deb/var/wazuh-agent_3.12.0-0.branchdev4466syscheckendpointimprovements.commit1a68474_amd64.deb)
  - [agent 3.12.0 0.branchdev4466syscheckendpointimprovements.commit1a68474 msi](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/windows/wazuh-agent-3.12.0-0.branchdev4466syscheckendpointimprovements.commit1a68474.msi)

### Stress test parameters:
- Manager system: `centos`
- Launch CentOS agent: `true`
- Launch Ubuntu agent: `true`
- Launch Windows agent: `true`
- Test duration: `60`
- Scan frequency: `10`
- Modules to test: 
  - logcollector: `true`
  - syscheck: `true`
  - rootcheck: `true`
  - sca: `true`
  - osquery: `true`
  - syscollector: `true`
  - active-response: `true`
  - azure-logs: `true`
  - cis-cat: `true`
  - docker-listener: `true`
  - open-scap: `true`
  - vulnerability-detector: `false`
- FIM special test: `None`
- FIM sleep: `300`
- FIM max files: `25000`
- SCA sleep: `30`
- SCA max registers: `5000`
- Docker image name: `hello-world`
- Docker sleep: `5`
- Docker jobs: `10`
- Logcollector special test: `None`
- Logcollector max files: `50000`
- Logcollector creation sleep: `300`
- Logcollector write files: `1000`
- Logcollector write sleep: `10`
- Vuln sleep: `60`
- Authd port: `1515`
- Syscollector package: `imagemagick.app`
- Syscollector package version: `7.0.8.48`

Regards!.

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind logcollector analysis:
**Commit:** 1a68474f44acdff09de054c638a8f8d969553e79
**Parameters:** track-origins=yes leak-check=full keep-stacktraces=alloc-and-free track-fds=yes num-callers=20 show-leak-kinds=definite,indirect
**Coverage:** 

```json
Lines executed:30.68% of 1832

```
**Coverage report detail:** [here](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.12.0/B180/coverage-report-detail)
## Valgrind result:
```
==15058== Memcheck, a memory error detector
==15058== Copyright (C) 2002-2017, and GNU GPLd, by Julian Seward et al.
==15058== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==15058== Command: /var/ossec/bin/ossec-logcollector -f
==15058== Parent PID: 1
==15058== 
==15058== 
==15058== FILE DESCRIPTORS: 17 open at exit.
==15058== Open file descriptor 15: /var/log/duplicated_file.log
==15058==    at 0x6221D19: open (open64.c:47)
==15058==    by 0x619E589: _IO_file_open (fileops.c:189)
==15058==    by 0x619E589: _IO_file_fopen@@GLIBC_2.2.5 (fileops.c:281)
==15058==    by 0x6190EA9: __fopen_internal (iofopen.c:78)
==15058==    by 0x6190EA9: fopen@@GLIBC_2.2.5 (iofopen.c:89)
==15058==    by 0x176217: reload_file (logcollector.c:903)
==15058==    by 0x1749F3: LogCollectorStart (logcollector.c:380)
==15058==    by 0x171E9A: main (main.c:187)
==15058== 
==15058== Open file descriptor 14: /var/log/target_file.log
==15058==    at 0x6221D19: open (open64.c:47)
==15058==    by 0x619E589: _IO_file_open (fileops.c:189)
==15058==    by 0x619E589: _IO_file_fopen@@GLIBC_2.2.5 (fileops.c:281)
==15058==    by 0x6190EA9: __fopen_internal (iofopen.c:78)
==15058==    by 0x6190EA9: fopen@@GLIBC_2.2.5 (iofopen.c:89)
==15058==    by 0x176217: reload_file (logcollector.c:903)
==15058==    by 0x1749F3: LogCollectorStart (logcollector.c:380)
==15058==    by 0x171E9A: main (main.c:187)
==15058== 
==15058== Open file descriptor 13: /var/log/label_file.log
==15058==    at 0x6221D19: open (open64.c:47)
==15058==    by 0x619E589: _IO_file_open (fileops.c:189)
==15058==    by 0x619E589: _IO_file_fopen@@GLIBC_2.2.5 (fileops.c:281)
==15058==    by 0x6190EA9: __fopen_internal (iofopen.c:78)
==15058==    by 0x6190EA9: fopen@@GLIBC_2.2.5 (iofopen.c:89)
==15058==    by 0x176217: reload_file (logcollector.c:903)
==15058==    by 0x1749F3: LogCollectorStart (logcollector.c:380)
==15058==    by 0x171E9A: main (main.c:187)
==15058== 
==15058== Open file descriptor 12: /var/log/json_file.log
==15058==    at 0x6221D19: open (open64.c:47)
==15058==    by 0x619E589: _IO_file_open (fileops.c:189)
==15058==    by 0x619E589: _IO_file_fopen@@GLIBC_2.2.5 (fileops.c:281)
==15058==    by 0x6190EA9: __fopen_internal (iofopen.c:78)
==15058==    by 0x6190EA9: fopen@@GLIBC_2.2.5 (iofopen.c:89)
==15058==    by 0x176217: reload_file (logcollector.c:903)
==15058==    by 0x1749F3: LogCollectorStart (logcollector.c:380)
==15058==    by 0x171E9A: main (main.c:187)
==15058== 
==15058== Open file descriptor 11: /var/log/alias_file.log
==15058==    at 0x6221D19: open (open64.c:47)
==15058==    by 0x619E589: _IO_file_open (fileops.c:189)
==15058==    by 0x619E589: _IO_file_fopen@@GLIBC_2.2.5 (fileops.c:281)
==15058==    by 0x6190EA9: __fopen_internal (iofopen.c:78)
==15058==    by 0x6190EA9: fopen@@GLIBC_2.2.5 (iofopen.c:89)
==15058==    by 0x176217: reload_file (logcollector.c:903)
==15058==    by 0x1749F3: LogCollectorStart (logcollector.c:380)
==15058==    by 0x171E9A: main (main.c:187)
==15058== 
==15058== Open file descriptor 10: /var/log/database_file.log-20-02-12.log
==15058==    at 0x6221D19: open (open64.c:47)
==15058==    by 0x619E589: _IO_file_open (fileops.c:189)
==15058==    by 0x619E589: _IO_file_fopen@@GLIBC_2.2.5 (fileops.c:281)
==15058==    by 0x6190EA9: __fopen_internal (iofopen.c:78)
==15058==    by 0x6190EA9: fopen@@GLIBC_2.2.5 (iofopen.c:89)
==15058==    by 0x176217: reload_file (logcollector.c:903)
==15058==    by 0x1749F3: LogCollectorStart (logcollector.c:380)
==15058==    by 0x171E9A: main (main.c:187)
==15058== 
==15058== Open file descriptor 9: /var/log/remove_file.log
==15058==    at 0x6221D19: open (open64.c:47)
==15058==    by 0x619E589: _IO_file_open (fileops.c:189)
==15058==    by 0x619E589: _IO_file_fopen@@GLIBC_2.2.5 (fileops.c:281)
==15058==    by 0x6190EA9: __fopen_internal (iofopen.c:78)
==15058==    by 0x6190EA9: fopen@@GLIBC_2.2.5 (iofopen.c:89)
==15058==    by 0x176217: reload_file (logcollector.c:903)
==15058==    by 0x1749F3: LogCollectorStart (logcollector.c:380)
==15058==    by 0x171E9A: main (main.c:187)
==15058== 
==15058== Open file descriptor 8: /var/log/single_file.log
==15058==    at 0x6221D19: open (open64.c:47)
==15058==    by 0x619E589: _IO_file_open (fileops.c:189)
==15058==    by 0x619E589: _IO_file_fopen@@GLIBC_2.2.5 (fileops.c:281)
==15058==    by 0x6190EA9: __fopen_internal (iofopen.c:78)
==15058==    by 0x6190EA9: fopen@@GLIBC_2.2.5 (iofopen.c:89)
==15058==    by 0x176217: reload_file (logcollector.c:903)
==15058==    by 0x1749F3: LogCollectorStart (logcollector.c:380)
==15058==    by 0x171E9A: main (main.c:187)
==15058== 
==15058== Open file descriptor 7: /var/log/dpkg.log
==15058==    at 0x6221D19: open (open64.c:47)
==15058==    by 0x619E589: _IO_file_open (fileops.c:189)
==15058==    by 0x619E589: _IO_file_fopen@@GLIBC_2.2.5 (fileops.c:281)
==15058==    by 0x6190EA9: __fopen_internal (iofopen.c:78)
==15058==    by 0x6190EA9: fopen@@GLIBC_2.2.5 (iofopen.c:89)
==15058==    by 0x176217: reload_file (logcollector.c:903)
==15058==    by 0x1749F3: LogCollectorStart (logcollector.c:380)
==15058==    by 0x171E9A: main (main.c:187)
==15058== 
==15058== Open file descriptor 6: /var/ossec/logs/active-responses.log
==15058==    at 0x6221D19: open (open64.c:47)
==15058==    by 0x619E589: _IO_file_open (fileops.c:189)
==15058==    by 0x619E589: _IO_file_fopen@@GLIBC_2.2.5 (fileops.c:281)
==15058==    by 0x6190EA9: __fopen_internal (iofopen.c:78)
==15058==    by 0x6190EA9: fopen@@GLIBC_2.2.5 (iofopen.c:89)
==15058==    by 0x176217: reload_file (logcollector.c:903)
==15058==    by 0x1749F3: LogCollectorStart (logcollector.c:380)
==15058==    by 0x171E9A: main (main.c:187)
==15058== 
==15058== Open AF_UNIX socket 16: /var/ossec/queue/ossec/logcollector
==15058==    at 0x6234EC7: socket (syscall-template.S:78)
==15058==    by 0x256040: OS_BindUnixDomain (os_net.c:153)
==15058==    by 0x16EFB4: lccom_main (lccom.c:102)
==15058==    by 0x5EFA6DA: start_thread (pthread_create.c:463)
==15058==    by 0x623388E: clone (clone.S:95)
==15058== 
==15058== Open AF_UNIX socket 5: <unknown>
==15058==    at 0x6234EC7: socket (syscall-template.S:78)
==15058==    by 0x256307: OS_ConnectUnixDomain (os_net.c:203)
==15058==    by 0x248024: StartMQ (mq_op.c:36)
==15058==    by 0x171E48: main (main.c:182)
==15058== 
==15058== Open file descriptor 4: /dev/urandom
==15058==    at 0x5F04D2B: open (open64.c:47)
==15058==    by 0x25230E: randombytes (randombytes.c:62)
==15058==    by 0x252440: srandom_init (randombytes.c:82)
==15058==    by 0x1717D6: main (main.c:60)
==15058== 
==15058== Open file descriptor 3: /tmp/dest/Analysis_dynamic_valgrind-v3.12.0-4586.log
==15058==    <inherited from parent>
==15058== 
==15058== Open file descriptor 2:
==15058==    <inherited from parent>
==15058== 
==15058== Open file descriptor 1:
==15058==    <inherited from parent>
==15058== 
==15058== Open file descriptor 0: /dev/null
==15058==    <inherited from parent>
==15058== 
==15058== 
==15058== HEAP SUMMARY:
==15058==     in use at exit: 56,319 bytes in 132 blocks
==15058==   total heap usage: 7,797 allocs, 7,665 frees, 4,605,849 bytes allocated
==15058== 
==15058== LEAK SUMMARY:
==15058==    definitely lost: 0 bytes in 0 blocks
==15058==    indirectly lost: 0 bytes in 0 blocks
==15058==      possibly lost: 1,904 bytes in 7 blocks
==15058==    still reachable: 54,415 bytes in 125 blocks
==15058==         suppressed: 0 bytes in 0 blocks
==15058== Reachable blocks (those to which a pointer was found) are not shown.
==15058== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==15058== 
==15058== For counts of detected and suppressed errors, rerun with: -v
==15058== ERROR SUMMARY: 3 errors from 3 contexts (suppressed: 0 from 0)

```

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind syscheckd analysis:
**Commit:** 1a68474f44acdff09de054c638a8f8d969553e79
**Parameters:** track-origins=yes leak-check=full keep-stacktraces=alloc-and-free track-fds=yes num-callers=20 show-leak-kinds=definite,indirect
**Coverage:** 

```json
Lines executed:25.18% of 2339

```
**Coverage report detail:** [here](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.12.0/B179/coverage-report-detail)
## Valgrind result:
```
==1527== Memcheck, a memory error detector
==1527== Copyright (C) 2002-2017, and GNU GPLd, by Julian Seward et al.
==1527== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==1527== Command: /var/ossec/bin/ossec-syscheckd -f
==1527== Parent PID: 15954
==1527== 
==1527== 
==1527== FILE DESCRIPTORS: 10 open at exit.
==1527== Open file descriptor 10: /usr/sbin/tcpdump
==1527==    at 0x6226D19: open (open64.c:47)
==1527==    by 0x61A3589: _IO_file_open (fileops.c:189)
==1527==    by 0x61A3589: _IO_file_fopen@@GLIBC_2.2.5 (fileops.c:281)
==1527==    by 0x6195EA9: __fopen_internal (iofopen.c:78)
==1527==    by 0x6195EA9: fopen@@GLIBC_2.2.5 (iofopen.c:89)
==1527==    by 0x194A34: os_string (os_string.c:169)
==1527==    by 0x18CE2B: check_rc_trojans (check_rc_trojans.c:91)
==1527==    by 0x18A8AB: run_rk_check (run_rk_check.c:140)
==1527==    by 0x18AF1F: w_rootcheck_thread (run_rk_check.c:319)
==1527==    by 0x5EFF6DA: start_thread (pthread_create.c:463)
==1527==    by 0x623888E: clone (clone.S:95)
==1527== 
==1527== Open file descriptor 9: /var/ossec/etc/rootcheck/rootkit_trojans.txt
==1527==    at 0x6226D19: open (open64.c:47)
==1527==    by 0x61A3589: _IO_file_open (fileops.c:189)
==1527==    by 0x61A3589: _IO_file_fopen@@GLIBC_2.2.5 (fileops.c:281)
==1527==    by 0x6195EA9: __fopen_internal (iofopen.c:78)
==1527==    by 0x6195EA9: fopen@@GLIBC_2.2.5 (iofopen.c:89)
==1527==    by 0x18A838: run_rk_check (run_rk_check.c:135)
==1527==    by 0x18AF1F: w_rootcheck_thread (run_rk_check.c:319)
==1527==    by 0x5EFF6DA: start_thread (pthread_create.c:463)
==1527==    by 0x623888E: clone (clone.S:95)
==1527== 
==1527== Open file descriptor 7: /dev/urandom
==1527==    at 0x5F09DAE: open (open64.c:47)
==1527==    by 0x265585: randombytes (randombytes.c:62)
==1527==    by 0x26571B: os_random (randombytes.c:88)
==1527==    by 0x25D71C: OSHash_Create (hash_op.c:54)
==1527==    by 0x17AFAA: fim_initialize (syscheck.c:88)
==1527==    by 0x17C06F: main (syscheck.c:472)
==1527== 
==1527== Open AF_UNIX socket 6: <unknown>
==1527==    at 0x6239EC7: socket (syscall-template.S:78)
==1527==    by 0x26957E: OS_ConnectUnixDomain (os_net.c:203)
==1527==    by 0x25C644: StartMQ (mq_op.c:36)
==1527==    by 0x17B90D: main (syscheck.c:396)
==1527== 
==1527== Open AF_UNIX socket 5: <unknown>
==1527==    at 0x6239EC7: socket (syscall-template.S:78)
==1527==    by 0x26957E: OS_ConnectUnixDomain (os_net.c:203)
==1527==    by 0x25C644: StartMQ (mq_op.c:36)
==1527==    by 0x189D26: rootcheck_connect (rootcheck.c:249)
==1527==    by 0x17B8DA: main (syscheck.c:389)
==1527== 
==1527== Open AF_UNIX socket 4: /var/ossec/queue/ossec/syscheck
==1527==    at 0x6239EC7: socket (syscall-template.S:78)
==1527==    by 0x2692B7: OS_BindUnixDomain (os_net.c:153)
==1527==    by 0x17E1B4: syscom_main (syscom.c:103)
==1527==    by 0x5EFF6DA: start_thread (pthread_create.c:463)
==1527==    by 0x623888E: clone (clone.S:95)
==1527== 
==1527== Open file descriptor 3: /tmp/dest/Analysis_dynamic_valgrind-v3.12.0-4586.log
==1527==    <inherited from parent>
==1527== 
==1527== Open file descriptor 2:
==1527==    <inherited from parent>
==1527== 
==1527== Open file descriptor 1:
==1527==    <inherited from parent>
==1527== 
==1527== Open file descriptor 0: /dev/null
==1527==    <inherited from parent>
==1527== 
==1527== 
==1527== HEAP SUMMARY:
==1527==     in use at exit: 130,672 bytes in 341 blocks
==1527==   total heap usage: 3,423,594 allocs, 3,423,253 frees, 460,705,752 bytes allocated
==1527== 
==1527== LEAK SUMMARY:
==1527==    definitely lost: 0 bytes in 0 blocks
==1527==    indirectly lost: 0 bytes in 0 blocks
==1527==      possibly lost: 816 bytes in 3 blocks
==1527==    still reachable: 129,856 bytes in 338 blocks
==1527==         suppressed: 0 bytes in 0 blocks
==1527== Reachable blocks (those to which a pointer was found) are not shown.
==1527== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==1527== 
==1527== For counts of detected and suppressed errors, rerun with: -v
==1527== ERROR SUMMARY: 3 errors from 3 contexts (suppressed: 0 from 0)

```

**Comment by wazuhci** [comment]
## Stress/Leaks test results
### Windows Agent ID0
- #### CPU
    [Windows Agent](https://ci.wazuh.com/data/Test_stress/warehouse-pullrequests/3.12.0/B128/monitor-winagent-ID0_B128_Test_stress_windows_Agent-warehouse-pullrequests_CPU(%25).svg)
- #### File Descriptors
    [Windows Agent](https://ci.wazuh.com/data/Test_stress/warehouse-pullrequests/3.12.0/B128/monitor-winagent-ID0_B128_Test_stress_windows_Agent-warehouse-pullrequests_FD.svg)
- #### Resident Set Size memory
    [Windows Agent](https://ci.wazuh.com/data/Test_stress/warehouse-pullrequests/3.12.0/B128/monitor-winagent-ID0_B128_Test_stress_windows_Agent-warehouse-pullrequests_RSS(KB).svg)
### CentOS Manager ID0
- #### CPU
    [CentOS Manager](https://ci.wazuh.com/data/Test_stress/warehouse-pullrequests/3.12.0/B128/monitor-manager-ID0_B128_Test_stress_Manager-warehouse-pullrequests_CPU(%25).svg)
- #### File Descriptors
    [CentOS Manager](https://ci.wazuh.com/data/Test_stress/warehouse-pullrequests/3.12.0/B128/monitor-manager-ID0_B128_Test_stress_Manager-warehouse-pullrequests_FD.svg)
- #### Resident Set Size memory
    [CentOS Manager](https://ci.wazuh.com/data/Test_stress/warehouse-pullrequests/3.12.0/B128/monitor-manager-ID0_B128_Test_stress_Manager-warehouse-pullrequests_RSS(KB).svg)
### CentOS Agent ID1
- #### CPU
    [CentOS Agent](https://ci.wazuh.com/data/Test_stress/warehouse-pullrequests/3.12.0/B128/monitor-agent-ID1_B128_Test_stress_centos_Agent-warehouse-pullrequests_CPU(%25).svg)
- #### File Descriptors
    [CentOS Agent](https://ci.wazuh.com/data/Test_stress/warehouse-pullrequests/3.12.0/B128/monitor-agent-ID1_B128_Test_stress_centos_Agent-warehouse-pullrequests_FD.svg)
- #### Resident Set Size memory
    [CentOS Agent](https://ci.wazuh.com/data/Test_stress/warehouse-pullrequests/3.12.0/B128/monitor-agent-ID1_B128_Test_stress_centos_Agent-warehouse-pullrequests_RSS(KB).svg)
### Ubuntu Agent ID0
- #### CPU
    [Ubuntu Agent](https://ci.wazuh.com/data/Test_stress/warehouse-pullrequests/3.12.0/B128/monitor-agent-ID0_B128_Test_stress_ubuntu_Agent-warehouse-pullrequests_CPU(%25).svg)
- #### File Descriptors
    [Ubuntu Agent](https://ci.wazuh.com/data/Test_stress/warehouse-pullrequests/3.12.0/B128/monitor-agent-ID0_B128_Test_stress_ubuntu_Agent-warehouse-pullrequests_FD.svg)
- #### Resident Set Size memory
    [Ubuntu Agent](https://ci.wazuh.com/data/Test_stress/warehouse-pullrequests/3.12.0/B128/monitor-agent-ID0_B128_Test_stress_ubuntu_Agent-warehouse-pullrequests_RSS(KB).svg)

File: framework/wazuh/utils.py [hunk 1]
```diff
@@ -962,9 +964,9 @@ def _parse_filters(self):
     def _process_filter(self, field_name, field_filter, q_filter):
         if field_name == "status":
             self._filter_status(q_filter)
-        elif field_name in self.date_fields and not self.date_regex.match(q_filter['value']):
-            # filter a date, but only if it is in timeframe format.
-            # If it matches the same format as DB (YYYY-MM-DD hh:mm:ss), filter directly by value (next if cond).
+        elif field_name in self.date_fields and not isinstance(q_filter['value'], int):
```
Comment (crd1985) [inline framework/wazuh/utils.py]:
```suggestion
        elif field_name in self.date_fields and not isinstance(q_filter['value'], (int, float)):
```

File: framework/wazuh/security_configuration_assessment.py [hunk 2]
```diff
@@ -65,9 +65,6 @@ def __init__(self, agent_id, offset, limit, sort, search, select, query, count,
     def _default_query(self):
         return self.default_query
 
-    def _default_count_query(self):
-        return f"COUNT(DISTINCT {self.count_field})"
-
```
Comment (crd1985) [inline framework/wazuh/security_configuration_assessment.py]: We should preserve this function but changing the count query as the super class does.

Comment (davidjiglesias) [reply framework/wazuh/security_configuration_assessment.py]: Added:
`return f"SELECT COUNT(DISTINCT {self.count_field})" + " FROM ({0})"`

**Pull request merged**